File: apps/frontend/src/app/foundation/vehicles/vehicle.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { ApiClient } from '@foundation/networking';
import { VehicleProfile, VehicleUpsertRequest } from './vehicle.models';

@Injectable({ providedIn: 'root' })
export class VehicleService {
  private readonly api = inject(ApiClient);

  listMyVehicles(): Observable<VehicleProfile[]> {
    return this.api.get<VehicleProfile[]>('/vehicles/me');
  }

  createVehicle(payload: VehicleUpsertRequest): Observable<VehicleProfile> {
    return this.api.post<VehicleProfile, VehicleUpsertRequest>('/vehicles', payload);
  }

  updateVehicle(vehicleId: string, payload: VehicleUpsertRequest): Observable<VehicleProfile> {
    return this.api.put<VehicleProfile, VehicleUpsertRequest>(
      `/vehicles/${encodeURIComponent(vehicleId)}`,
      payload,
    );
  }

  deleteVehicle(vehicleId: string): Observable<void> {
    return this.api.delete<void>(`/vehicles/${encodeURIComponent(vehicleId)}`);
  }
}

File: apps/ios/SpotLink/Sources/SpotLink/Features/Vehicles/VehicleService.swift
import Foundation

// MARK: - Vehicle Service

public final class VehicleService: Sendable {
    private let apiClient: APIClientProtocol

    public init(apiClient: APIClientProtocol) {
        self.apiClient = apiClient
    }

    public func listMyVehicles() async throws -> [VehicleProfile] {
        try await apiClient.get("/vehicles/me", query: nil)
    }

    public func createVehicle(_ request: VehicleUpsertRequest) async throws -> VehicleProfile {
        try await apiClient.post("/vehicles", body: request)
    }

    public func updateVehicle(_ vehicleId: String, request: VehicleUpsertRequest) async throws -> VehicleProfile {
        try await apiClient.put("/vehicles/\(vehicleId)", body: request)
    }

    public func deleteVehicle(_ vehicleId: String) async throws {
        try await apiClient.delete("/vehicles/\(vehicleId)")
    }
}
